primitive/zk_rollup/src/lib.rs

```diff
@@ -11,47 +11,81 @@ mod verifier_contract;
 #[cfg(test)]
 mod tests {
 
-    use jub_jub::Fp;
+    use std::vec;
+
+    use jub_jub::{Fp, JubjubExtended};
+    use rand::rngs::StdRng;
+    use rand_core::SeedableRng;
+    use red_jubjub::{PublicKey, SecretKey};
+    use zkstd::common::{CurveGroup, Group};
 
     use crate::{
         main_contract::MainContract,
-        merkle_tree::MerkleProof,
-        operator::{RollupOperator, Transaction, UserData},
+        operator::{RollupOperator, TransactionData},
         poseidon::Poseidon,
     };
 
     #[test]
-    #[should_panic]
     fn test_deposit_and_transfer_works() {
-        let mut contract = MainContract::<Fp>::default();
+        let mut rng = StdRng::seed_from_u64(8349u64);
+        let mut contract =
+            MainContract::<Fp>::new(PublicKey::new(JubjubExtended::random(&mut rng)));
+        let mut operator = RollupOperator::<Fp, Poseidon<Fp, 2>, ACCOUNT_LIMIT>::default();
         let poseidon = Poseidon::<Fp, 2>::new();
-        const ACCOUNT_LIMIT: usize = 3;
+        const ACCOUNT_LIMIT: usize = 2;
 
-        // deploy contract
+        let alice_secret = SecretKey::new(Fp::random(&mut rng));
+        let bob_secret = SecretKey::new(Fp::random(&mut rng));
+        let alice_address = alice_secret.to_public_key();
+        let bob_address = bob_secret.to_public_key();
 
-        let user = UserData::default();
-        contract.deposit(10);
+        let root_before_dep = operator.state_root();
+        assert_eq!(root_before_dep, Fp::zero());
 
-        // Set BATCH_SIZE to 1
+        let deposit1 = TransactionData::new(alice_address, contract.address(), 10)
+            .signed(alice_secret, &mut rng);
+        let deposit2 =
+            TransactionData::new(bob_address, contract.address(), 0).signed(bob_secret, &mut rng);
+        contract.deposit(deposit1);
+        contract.deposit(deposit2);
 
-        let mut operator = RollupOperator::<Fp, Poseidon<Fp, 2>, ACCOUNT_LIMIT>::default();
+        let pending_deposits = contract.deposits();
+        assert_eq!(pending_deposits.len(), ACCOUNT_LIMIT);
+        operator.process_deposits(pending_deposits.clone(), &poseidon);
+
+        let root_after_dep = operator.state_root();
+
+        assert_eq!(
+            root_after_dep,
+            Fp::from_hex("0x0b110370e1cda66a7105997fe1196ffe2a685a8d6ccba7aad79e21823cc163c0")
+                .unwrap()
+        );
 
-        // operator.run();
         // should periodically check deposits on the contract
 
-        assert!(contract.check_balance(MerkleProof::default()) == user.balance());
+        // assert!(contract.check_balance(MerkleProof::default()) == alice.balance());
```

Comment:
operator needs to init database Alice and Bob balance
Alice and Bob query to database